ta/pkcs11/src/object.c

```diff
@@ -902,3 +902,276 @@ uint32_t entry_get_object_size(struct pkcs11_client *client,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	TEE_Param *out = params + 2;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct pkcs11_object_head *template = NULL;
+	struct obj_attrs *head = NULL;
+	struct obj_attrs *head_new = NULL;
+	size_t template_size = 0;
+	struct pkcs11_object *obj = NULL;
+	uint32_t object_handle = 0;
+	uint32_t obj_handle = 0;
+	enum processing_func function = PKCS11_FUNCTION_COPY;
+	enum pkcs11_class_id class = PKCS11_CKO_UNDEFINED_ID;
+
+	if (!client || ptypes != exp_pt ||
+	    out->memref.size != sizeof(obj_handle))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &object_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_alloc_get_attributes(&ctrlargs, &template);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	obj = pkcs11_handle2object(object_handle, session);
+	if (!obj) {
+		rc = PKCS11_CKR_OBJECT_HANDLE_INVALID;
+		goto out;
+	}
+
+	/* Only session objects can be modified during a read-only session */
+	if (object_is_token(obj->attributes) &&
+	    !pkcs11_session_is_read_write(session)) {
+		DMSG("Can't modify persistent object in a RO session");
+		rc = PKCS11_CKR_SESSION_READ_ONLY;
+		goto out;
+	}
+
+	/*
+	 * Only public objects can be modified unless normal user is logged in
+	 */
+	rc = check_access_attrs_against_token(session, obj->attributes);
+	if (rc) {
+		rc = PKCS11_CKR_USER_NOT_LOGGED_IN;
+		goto out;
+	}
+
+	/* Objects with PKCS11_CKA_COPYABLE as false aren't modifiable */
+	if (!object_is_copyable(obj->attributes)) {
+		rc = PKCS11_CKR_ACTION_PROHIBITED;
+		goto out;
+	}
+
+	template_size = sizeof(*template) + template->attrs_size;
+
+	/*
+	 * Prepare a clean initial state (@head) for the template. Helps in
+	 * removing any duplicates or inconsistent values from the
+	 * template.
+	 */
+	rc = create_attributes_from_template(&head, template, template_size,
+					     NULL, function,
+					     PKCS11_CKM_UNDEFINED_ID,
+					     PKCS11_CKO_UNDEFINED_ID);
+	if (rc)
+		goto out;
+
+	/* Check the attributes in @head to see if they are modifiable */
+	rc = check_attrs_against_modification(session, head, obj, function);
+	if (rc)
+		goto out;
+
+	class = get_class(obj->attributes);
+
+	if (class == PKCS11_CKO_SECRET_KEY ||
+	    class == PKCS11_CKO_PRIVATE_KEY) {
+		/*
+		 * If CKA_EXTRACTABLE attribute in passed template (@head) is
+		 * modified to CKA_FALSE, CKA_NEVER_EXTRACTABLE should also
+		 * change to CKA_FALSE in copied obj. So, add it to the
+		 * passed template.
+		 */
+		uint8_t bbool = 0;
+		uint32_t size = sizeof(bbool);
+
+		rc = get_attribute(head, PKCS11_CKA_EXTRACTABLE, &bbool, &size);
```

Comment:
Don't we need similar treatment for `CKA_SENSITIVE` and `CKA_ALWAYS_SENSITIVE`?

Reply:
Special treatment is not required for CKA_SENSITIVE as in this case the copies object inherits the value of CKA_ALWAYS_SENSITIVE from parent.

PKCS#11 v2.40 spec - in C_CopyObject Description -
The template may specify new values for any attributes of the object that can ordinarily be modified (e.g.,
in the course of copying a secret key, a key’s CKA_EXTRACTABLE attribute may be changed from
CK_TRUE to CK_FALSE, but not the other way around. If this change is made, the new key’s
CKA_NEVER_EXTRACTABLE attribute will have the value CK_FALSE. Similarly, the template may
specify that the new key’s CKA_SENSITIVE attribute be CK_TRUE; the new key will have the same
value for its CKA_ALWAYS_SENSITIVE attribute as the original key).

Reply:
Ideally this check should not be required as per the explanation I added in C_SetAttributeValue() comment below. However, since specification explicitly mentioned it for C_CopyObject(), I added the check just to avoid any corner scenario we may be missing.